<a href="https://colab.research.google.com/github/adong-hood/cs200/blob/main/assignment7_web_scrapping_II.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Web Scraping Practice II

The CIA website stopped publishing the world factbook in 2026 (<a href = "https://www.cia.gov/stories/story/spotlighting-the-world-factbook-as-we-bid-a-fond-farewell/">Spotlighting The World Factbook as We Bid a Fond Farewell
 </a>). In this practice, we will use the snapshot of August 06, 2025 from the web archive. This page--[Guide to Country Profile](https://web.archive.org/web/20260204071503/https://www.cia.gov/the-world-factbook/countries/world/)--is very similar to the [Notes and Definitions](http://pluto.hood.edu/~dong/datasets/factbook/docs/notesanddefs.html) page from the 2017 version.

To make the problem easier, we will let you start with this page directly: [Infant Mortality Rate](https://web.archive.org/web/20260203121119/https://www.cia.gov/the-world-factbook/field/infant-mortality-rate/) page directly.


**Your Tasks**

Remember to share your notebook for any credits.
<pre>
Step 1: scrape the total infant mortality rates from 2024 (est.).
Step 2: scrape the total infant mortality rates from 2017(refer to assignment 6).
Step 3: merge the infant mortality rates of 2017 with those of 2024.
Find which country has the largest positive infant mortality rate change from 2017 to 2024 and which country has the largest negative infant mortality rate change during the same time period.
</pre>


In [ ]:
import pandas as pd
import numpy as np
import altair as alt
import requests
from bs4 import BeautifulSoup

## Step 1. Scrape the total infant mortality rates for 2024 (estimated).

In [ ]:
      # <div class="pb30">
      #   <h3 class="mt10"><a href="/countries/afghanistan/">Afghanistan</a></h3>
      #   <p><strong>total:</strong> 101.3 deaths/1,000 live births (2024 est.)</p>
      #   <p><strong>male:</strong> 109.7 deaths/1,000 live births</p>
      #   <p><strong>female:</strong> 92.5 deaths/1,000 live births</p>
      # </div>


In [ ]:
# This web archive link is slow and not always reliable.
#address = 'https://web.archive.org/web/20260203121119/https://www.cia.gov/the-world-factbook/field/infant-mortality-rate/'
#- Insteadn, use this link for 2024 rates.
address_2024 = 'http://pluto.hood.edu/~dong/datasets/factbook/fields/infantmortality-2024.html'
res = requests.get(address_2024)
print(res.status_code)
page_2024 = res.text
soup_2024 = BeautifulSoup(page_2024)
all_countries = soup_2024.select('.pb30')
infant_mortality_df_2024 = {}
for country in all_countries:
  country_name = country.find('h3').get_text()
  total_rate = country.find('p').text.strip().split(":")[1].strip().split('deaths/')[0].strip()
  # print(country_name)
  # print(total_rate)
  infant_mortality_df_2024[country_name] = float(total_rate)

infant_mortality_df_2024
infant_2024_df = pd.DataFrame({'infant-mortality-rate-2024':infant_mortality_df_2024})
infant_2024_df.head()

200


,infant-mortality-rate-2024
Afghanistan,101.3
Albania,10.3
Algeria,18.7
American Samoa,9.7
Andorra,3.3


In [ ]:
     # <div class="pb30">
      #   <h3 class="mt10"><a href="/countries/albania/">Albania</a></h3>
      #   <p><strong>total:</strong> 10.3 deaths/1,000 live births (2024 est.)</p>
      #   <p><strong>male:</strong> 11.3 deaths/1,000 live births</p>
      #   <p><strong>female:</strong> 9.2 deaths/1,000 live births</p>
      # </div>

In [ ]:
#- For checking after all rates are extracted.
infant_mortality_df_2024["infant-mortality-rate-2024"].mean()

np.float64(18.67467248908297)

## Step 2. Scrape the total infant mortality rates for 2017 (assignment 6).

In [ ]:
#Import over all code from sssignment 6 for the 2017 data
address_2017 = 'http://pluto.hood.edu/~dong/datasets/factbook/fields/2091.html#118'
res = requests.get(address_2017)
page_2017 = res.text

# fill in the code for the extraction.

## Step 3.1 Combine and sort the data.

## Step 3.2 Find the country with the largest positive change.

## Step 3.3 Find the country with the largest negative change.